src/main/java/jenkins/scm/api/SCMEvent.java

```diff
@@ -200,10 +206,13 @@ protected SCMEvent(SCMEvent<P> copy) {
      * @return a {@link ScheduledExecutorService}.
      */
     @NonNull
-    protected static ScheduledExecutorService executorService() {
-        // Future-proofing, if we find out that events are drowning Timer then we may need to move them to their
-        // own dedicated ScheduledExecutorService thread pool
-        return Timer.get();
+    protected static synchronized ScheduledExecutorService executorService() {
+        if (executorService == null) {
+            // corePoolSize is set to 10, but will only be created if needed.
+            // ScheduledThreadPoolExecutor "acts as a fixed-sized pool using corePoolSize threads"
+            executorService =  new ScheduledThreadPoolExecutor(10, new NamingThreadFactory(new DaemonThreadFactory(), "SCMEvent"));
```

Comment:
Note that `Timer` is also wrapped in utilities to

* impersonate `ACL.SYSTEM`
* handle errors, if we are not using `SafeTimerTask`
* maintain the `Thread.contextClassLoader`

Whether any of those issues are important for this usage, I have no idea offhand. But perhaps `Timer` should get a utility factory

```java
public static ScheduledExecutorService newService(String name, int corePoolSize) {
    return new ImpersonatingScheduledExecutorService(new ErrorLoggingScheduledThreadPoolExecutor(corePoolSize, new NamingThreadFactory(new ClassLoaderSanityThreadFactory(new DaemonThreadFactory()), name)), ACL.SYSTEM);
}
```

for easy reuse.

Reply:
This one does everything except logging errors if we are not using `SafeTimerTask`

I can use exactly the same as in timer... but I choose to use that one for some purpose..
It was so long ago I dont even remember why.. 
But I remember it was not accident..

Reply:
And `Timer` is in core so I cannot change it from here..

Reply:
> This one does everything except logging errors

It does not do the impersonation or context class loader maintenance.

> `Timer` is in core so I cannot change it from here

Of course, but we can have a core PR, and via [JEP-305](https://jenkins.io/jep/305) it is even straightforward to have an `on-hold` PR in the plugin which verifies its usage.

Reply:
I see that there was change in Timer.. ok.. so it does more now..

So it would be great to move it somewhere in one place..

And maybe it would be easier and better to just allow changing size of timer pool via variable or something?
Here https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/util/Timer.java#L47

Reply:
Again, to be clear, I am not _requesting_ that the executor service here add all of those protections, I am just noting for purposes of review that it does not.

> maybe it would be easier and better to just allow changing size of timer pool

Perhaps, but that is independent of this issue. The `Timer` pool is intended to be used for quick, nonblocking tasks. The implication of this PR is that some SCM event handlers were not quick. There is no associated JIRA issue or example thread dump so I cannot evaluate whether they were _intended_ to be quick and are not due to some bug, or whether it is just natural that they sometimes block for a while and thus should never have been using `Timer` to begin with.

Reply:
They are not quick.. they sometimes takes hours to complete.. (partially because of work that they need to do or of other bugs that I will try to talk about maybe on gitter, I need to ask how we would like to resolve them)

So pool size in timer is irrelevant in this context..

And about that added protections Im not sure also.. this local pool does one rather specified task, so maybe they are not needed.
But to be sure I would like to add them here as they are in Timer at current moment.

Reply:
> they sometimes takes hours to complete

A thread dump would be the background we need to analyze why.

Reply:
I can get thread dump.. just need to wait for busy day..
But all waits here https://github.com/kohsuke/github-api/blob/master/src/main/java/org/kohsuke/github/RateLimitHandler.java#L39
We are just got rate limit badly..

Reply:
And while we dont have problems with waiting for GH.. 
We have problem that nothing else (that relies on `Timer`) on jenkins is working..

Reply:
Need to thread dump to see what _caller_ is to blame.

Reply:
Webhook processing is going into this pool..
And there is issue with GH processing itself (sometimes rate limiting is not resetting)..
But not that easy to grasp..
And would be best to fix somewhere here..
https://github.com/kohsuke/github-api/blob/master/src/main/java/org/kohsuke/github/GitHub.java#L316
Once for good :)